### DISCUSSION TITLE
Can I horizontally scale a discord.js bot?

### DISCUSSION THREAD

**Discussion by heyAyushh** [discordjs/discord.js#5780] [Q&A] [answered]
Right now we have a express server where we also included discord.js code, 
If we tried to scale the bot horizontally multiple instances of same client starts running,

the problem is when there's a   
```
client.on('guildMemberAdd', member => welcomeMember({ member }));
```  
Multiple welcome messages are being sent for only one member joining the server,  
it might happen to other event listeners as well.
  
What's the solution for this?

The event guildMemberAdd should be delivered exactly once irrespective on number of clients that i have.

**Comment by kyranet** [comment, marked as answer]
To horizontally scale a Discord bot within a server, you need to use a Sharding Manager, more information in [the guide](https://discordjs.guide/sharding/#sharding-file).

As for scaling it across different servers... that's a lot trickier and I don't know a sharder that does this.

---

Not really related to the issue at hand but... within the same server, you can alternative do any of the following:

- Separate the express server from the bot, and make them communicate in an alternative way.
- Use a much more scalable HTTP framework, [`fastify`](https://www.npmjs.com/package/fastify) has a plugin called [`fastify-piscina`](https://www.npmjs.com/package/fastify-piscina) which supports worker threads, which are more efficient than separate processes as they share information quicker and share the same memory.

Generally speaking, discord.js is fine in a single process even if you're at 10k guilds, so if your bot is below that, you should not need to do this. Also, it's not recommended to shard below 1k guilds as it wastes your daily `IDENTIFY`s and creates a large waste of resources.

**Reply by heyAyushh** [reply]
I am talking about scaling this to different servers not because I don't want to scale the bot, but to scale the express server, or add a load balancer.

guess, i'll have to separate the bot code from express code.

I was looking for something that will give me exactly once delivery of the event.

**Reply by kyranet** [reply]
Although very advanced, there are libraries that allow you to define brokers in both a centralized and a de-centralized manner, some supporting kubernetes, however, you wouldn't be able to use discord.js with them due to a number of issues (such as shared cache and ratelimit data sharing).

So yes, the best solution you have at hand right now, is to separate both codebases.

**Reply by heyAyushh** [reply]
Thankyou so much @kyranet
